src/net/ip4/icmp4.cpp

```diff
@@ -0,0 +1,234 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// #define DEBUG
+#include <net/ip4/icmp4.hpp>
+
+namespace net {
+
+  // ---------------------------- ICMP_packet ----------------------------
+
+  std::string ICMP_packet::to_string() {
+    if (type_ == icmp4::Type::NO_REPLY)
+      return "No reply received";
+
+    return "Identifier: " + std::to_string(id_) + "\n" +
+      "Sequence number: " + std::to_string(seq_) + "\n" +
+      "Source: " + src_.to_string() + "\n" +
+      "Destination: " + dst_.to_string() + "\n" +
+      "Type: " + icmp4::get_type_string(type_) + "\n" +
+      "Code: " + icmp4::get_code_string(type_, code_) + "\n" +
+      "Checksum: " + std::to_string(checksum_) + "\n" +
+      "Data: " + std::string{payload_.begin(), payload_.end()};
+  }
+
+  // ------------------------------ ICMPv4 ------------------------------
+
+  int ICMPv4::request_id_ = 0;
+
+  ICMPv4::ICMPv4(Stack& inet) :
+    inet_{inet}
+  {}
+
+  void ICMPv4::receive(Packet_ptr pckt) {
+    if ((size_t)pckt->size() < sizeof(IP4::header) + icmp4::Packet::header_size()) // Drop if not a full header
+      return;
+
+    auto pckt_ip4 = static_unique_ptr_cast<PacketIP4>(std::move(pckt));
+    auto req = icmp4::Packet(std::move(pckt_ip4));
+
+    switch(req.type()) {
+    case (icmp4::Type::ECHO):
+      debug("<ICMP> PING from %s\n", req.ip().src().str().c_str());
+      ping_reply(req);
+      break;
+    case (icmp4::Type::ECHO_REPLY):
+      debug("<ICMP> PING Reply from %s\n", req.ip().src().str().c_str());
+      execute_ping_callback(req);
+      break;
+    case (icmp4::Type::DEST_UNREACHABLE):
+    case (icmp4::Type::REDIRECT):
+    case (icmp4::Type::TIME_EXCEEDED):
+    case (icmp4::Type::PARAMETER_PROBLEM):
+      debug("<ICMP> ICMP error message from %s\n", req.ip().src().str().c_str());
+      forward_to_transport_layer(req);
+      break;
+    case (icmp4::Type::TIMESTAMP):
+      debug("<ICMP> TIMESTAMP from %s\n", req.ip().src().str().c_str());
+      // TODO May
+      // timestamp_reply(req);
+      break;
+    case (icmp4::Type::TIMESTAMP_REPLY):
+      debug("<ICMP> TIMESTAMP REPLY from %s\n", req.ip().src().str().c_str());
+      // TODO May
+      break;
+    default:  // icmp4::NO_REPLY
+      return;
+    }
+  }
+
+  void ICMPv4::forward_to_transport_layer(icmp4::Packet& req) {
+    // inet forwards to transport layer (UDP or TCP)
+    inet_.error_report(req.type(), req.code(), req.payload());
+  }
+
+  void ICMPv4::destination_unreachable(Packet_ptr pckt, icmp4::code::Dest_unreachable code) {
+    if ((size_t)pckt->size() < sizeof(IP4::header) + icmp4::Packet::header_size()) // Drop if not a full header
```

Comment:
You could make it more DRY by creating a inline helper function for this, and also use UNLIKELY:
`if(UNLIKELY(not full_header(*pckt))) return;`
(Could this be done before any of the functions are called?)